### ISSUE TITLE
pts.ply and camera poses is working but colmap points and splat GSs are with some problems

### ISSUE THREAD

**Issue by a11enL** [Tencent-Hunyuan/HunyuanWorld-Mirror#21]
Thank for your great work.  I need help to correct the colmap 3D points and Gaussian GSs. currently for my testing with infer.py script:

pts.ply is working with correct position
camera poses with colmap are working with correct position and direction.

but the colmap 3D points and Gaussian GSs are with wrong position or rotation as below screenshot.

<img width="813" height="621" alt="Image" src="https://github.com/user-attachments/assets/c045ca6b-bb16-44bd-9453-38f80b033ecb" />

the correct pts.ply show as below：

<img width="717" height="551" alt="Image" src="https://github.com/user-attachments/assets/91ecd618-eb39-48b2-8c4c-629092f0de90" />

I test with two datasets with 12 images and 100 images, both of them are with same result as above.

**Comment by zhenwwang** [comment]
Hi @a11enL, what's the 3dgs viewer you used? The gaussian.ply looks fine in my side with a supersplat viewer (https://superspl.at/editor/) or https://antimatter15.com/splat/

**Comment by a11enL** [comment]
Hi @zhenwwang , I use my custom Viser-based 3dgs viewer, I also tested it with Meshlab.

I have a question for the data in predictions["splats"]["means"], is it in camera space or world space? I transform it to world space with the first view's camera coordinates coordinates, but still with wrong result. by the way, all data are correct just with a bit problem with gaussian.ply.

    if "splats" in predictions and args.save_gs and True:
        means = predictions["splats"]["means"][0].reshape(-1, 3)
        scales = predictions["splats"]["scales"][0].reshape(-1, 3)
        quats = predictions["splats"]["quats"][0].reshape(-1, 4)
        colors = (predictions["splats"]["sh"][0] if "sh" in predictions["splats"] else predictions["splats"]["colors"][0]).reshape(-1, 3)
        opacities = predictions["splats"]["opacities"][0].reshape(-1)

        # Ensure Gaussian means are in the correct world coordinate system
        if "camera_params" in predictions:
            e3x4, _ = vector_to_camera_matrices(predictions["camera_params"], image_hw=(H, W))
            w2c = e3x4[0, 0, :3, :4]  # [3, 4] world-to-camera for first view
            w2c_homo = torch.cat([w2c, torch.tensor([[0, 0, 0, 1]], dtype=w2c.dtype, device=w2c.device)], dim=0)  # [4, 4]
            if w2c_homo.shape != (4, 4):
                raise ValueError(f"Expected w2c_homo shape [4, 4], got {w2c_homo.shape}")
            c2w = torch.linalg.inv(w2c_homo)  # [4, 4] camera-to-world
            # Transform means to world coordinates (assuming means are in camera coordinates)
            means_homo = torch.cat([means, torch.ones_like(means[:, :1])], dim=1)  # [N, 4]
            means_world = torch.matmul(means_homo, c2w.T)[:, :3]  # [N, 3]
            means = means_world
        else:
            print("⚠️ No camera_params found, assuming means are already in world coordinates")

        # Ensure tensors are on CPU and in float32 for saving
        means = means.float().detach().cpu()
        scales = scales.float().detach().cpu()
        quats = quats.float().detach().cpu()
        colors = colors.float().detach().cpu()
        opacities = opacities.float().detach().cpu()


> Hi [@a11enL](https://github.com/a11enL), what's the 3dgs viewer you used? The gaussian.ply looks fine in my side with a supersplat viewer (https://superspl.at/editor/) or https://antimatter15.com/splat/

**Comment by yifliu3** [comment]
> Hi [@zhenwwang](https://github.com/zhenwwang) , I use my custom Viser-based 3dgs viewer, I also tested it with Meshlab.
> 
> I have a question for the data in predictions["splats"]["means"], is it in camera space or world space? I transform it to world space with the first view's camera coordinates coordinates, but still with wrong result. by the way, all data are correct just with a bit problem with gaussian.ply.
> 
> ```
> if "splats" in predictions and args.save_gs and True:
>     means = predictions["splats"]["means"][0].reshape(-1, 3)
>     scales = predictions["splats"]["scales"][0].reshape(-1, 3)
>     quats = predictions["splats"]["quats"][0].reshape(-1, 4)
>     colors = (predictions["splats"]["sh"][0] if "sh" in predictions["splats"] else predictions["splats"]["colors"][0]).reshape(-1, 3)
>     opacities = predictions["splats"]["opacities"][0].reshape(-1)
> 
>     # Ensure Gaussian means are in the correct world coordinate system
>     if "camera_params" in predictions:
>         e3x4, _ = vector_to_camera_matrices(predictions["camera_params"], image_hw=(H, W))
>         w2c = e3x4[0, 0, :3, :4]  # [3, 4] world-to-camera for first view
>         w2c_homo = torch.cat([w2c, torch.tensor([[0, 0, 0, 1]], dtype=w2c.dtype, device=w2c.device)], dim=0)  # [4, 4]
>         if w2c_homo.shape != (4, 4):
>             raise ValueError(f"Expected w2c_homo shape [4, 4], got {w2c_homo.shape}")
>         c2w = torch.linalg.inv(w2c_homo)  # [4, 4] camera-to-world
>         # Transform means to world coordinates (assuming means are in camera coordinates)
>         means_homo = torch.cat([means, torch.ones_like(means[:, :1])], dim=1)  # [N, 4]
>         means_world = torch.matmul(means_homo, c2w.T)[:, :3]  # [N, 3]
>         means = means_world
>     else:
>         print("⚠️ No camera_params found, assuming means are already in world coordinates")
> 
>     # Ensure tensors are on CPU and in float32 for saving
>     means = means.float().detach().cpu()
>     scales = scales.float().detach().cpu()
>     quats = quats.float().detach().cpu()
>     colors = colors.float().detach().cpu()
>     opacities = opacities.float().detach().cpu()
> ```
> 
> > Hi [@a11enL](https://github.com/a11enL), what's the 3dgs viewer you used? The gaussian.ply looks fine in my side with a supersplat viewer (https://superspl.at/editor/) or https://antimatter15.com/splat/

Dear dev, the means of the 3D Gaussians are defined in world space, which is anchored to the coordinate system of the first camera. We’ve tested several cases and found that the visualizations appear correct. If you’re still encountering issues, please try using the latest version of the code or test with additional scenes.
